substrate/frame/parameters/src/lib.rs

```diff
@@ -0,0 +1,252 @@
+#![doc = docify::embed!("src/tests/unit.rs", set_parameters_example)]
+//!
+//! ## Low Level / Implementation Details
+//!
+//! The pallet stores the parameters in a storage map and implements the matching `Get<Value>` for
+//! each `Key` type. The `Get` then accesses the `Parameters` map to retrieve the value. An event is
+//! emitted every time that a value was updated. It is even emitted when the value is changed to the
+//! same.
+//!
+//! The key and value types themselves are defined by macros and aggregated into a runtime wide
+//! enum. This enum is then injected into the pallet. This allows it to be used without any changed
+//! to the pallet that the parameter will be utilized by.
+//!
+//! ### Design Goals
+//!
+//! 1. Easy to update without runtime upgrade.
+//! 2. Exposes metadata and docs for user convenience.
+//! 3. Can be permissioned on a per-key base.
+//!
+//! ### Design
+//!
+//! 1. Everything is done at runtime without the need for `const` values. `Get` allows for this -
+//! which is coincidentally an upside and a downside. 2. The types are defined through macros, which
+//! allows to expose metadata and docs. 3. Access control is done through the `EnsureOriginWithArg`
+//! trait, that allows to pass data along to the origin check. It gets passed in the key. The
+//! implementor can then match on the key and the origin to decide whether the origin is
+//! permissioned to set the value.
+
+use frame_support::pallet_prelude::*;
+use frame_system::pallet_prelude::*;
+
+use frame_support::traits::{
+	dynamic_params::{AggregratedKeyValue, IntoKey, Key, RuntimeParameterStore, TryIntoKey},
+	EnsureOriginWithArg,
+};
+
+mod benchmarking;
+#[cfg(test)]
+mod tests;
+mod weights;
+
+pub use pallet::*;
+pub use weights::WeightInfo;
+
+/// The key type of a parameter.
+type KeyOf<T> = <<T as Config>::RuntimeParameters as AggregratedKeyValue>::AggregratedKey;
+
+/// The value type of a parameter.
+type ValueOf<T> = <<T as Config>::RuntimeParameters as AggregratedKeyValue>::AggregratedValue;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+
+	#[pallet::config(with_default)]
+	pub trait Config: frame_system::Config {
+		/// The overarching event type.
+		#[pallet::no_default_bounds]
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+
+		/// The overarching KV type of the parameters.
+		///
+		/// Usually created by [`frame_support::dynamic_params`] or equivalent.
+		#[pallet::no_default_bounds]
+		type RuntimeParameters: AggregratedKeyValue;
+
+		/// The origin which may update a parameter.
+		///
+		/// The key of the parameter is passed in as second argument to allow for fine grained
+		/// control.
+		#[pallet::no_default_bounds]
+		type AdminOrigin: EnsureOriginWithArg<Self::RuntimeOrigin, KeyOf<Self>>;
+
+		/// Weight information for extrinsics in this module.
+		type WeightInfo: WeightInfo;
+
+		/// Provides a default KV since the type is otherwise in-constructable.
```

Comment:
> in-constructable

What does this mean?

Reply:
It means that its impossible to create an instance of it. Basically i want `RuntimeParameters: Default`, but only if the benchmarking is enabled. Since a default does not make any sense otherwise.  
This can also be generated by the macro eventually, for now its manual.